src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+    VerifyOrReturnValue(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), TimeState::kInvalid);
+
+    for (size_t i = 0; i < tzList.size(); i++)
+    {
+        auto & tz = tzList[i].timeZone;
+        if (tz.validAt != 0 && tz.validAt <= chipEpochTime)
+        {
+            tz.validAt    = 0;
+            activeTzIndex = i;
+        }
+    }
+    if (activeTzIndex != 0)
+    {
+        mTimeZoneObj.size    = tzList.size() - activeTzIndex;
+        auto newTimeZoneList = tzList.SubSpan(activeTzIndex);
+        VerifyOrReturnValue(mTimeSyncDataProvider.StoreTimeZone(newTimeZoneList) == CHIP_NO_ERROR, TimeState::kInvalid);
+        VerifyOrReturnValue(LoadTimeZone() == CHIP_NO_ERROR, TimeState::kInvalid);
+        return TimeState::kChanged;
+    }
+    return TimeState::kActive;
+}
+
+TimeState TimeSynchronizationServer::GetUpdatedDSTOffsetState()
+{
+    System::Clock::Microseconds64 utcTime;
+    size_t activeDstIndex = 0;
+    auto dstList          = GetDSTOffset();
+    uint64_t chipEpochTime;
+    bool dstStopped = true;
+
+    VerifyOrReturnValue(System::SystemClock().GetClock_RealTime(utcTime) == CHIP_NO_ERROR, TimeState::kInvalid);
+    VerifyOrReturnValue(dstList.size() != 0, TimeState::kInvalid);
+    VerifyOrReturnValue(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), TimeState::kInvalid);
+
+    for (size_t i = 0; i < dstList.size(); i++)
+    {
+        if (dstList[i].validStarting <= chipEpochTime)
+        {
+            activeDstIndex = i;
+            dstStopped     = false;
+        }
+    }
+    VerifyOrReturnValue(dstStopped, TimeState::kStopped);
+    // if offset is zero and validUntil is null then no DST is used
+    if (dstList[activeDstIndex].offset == 0 && dstList[activeDstIndex].validUntil.IsNull())
+    {
+        return TimeState::kStopped;
+    }
+    if (!dstList[activeDstIndex].validUntil.IsNull() && dstList[activeDstIndex].validUntil.Value() <= chipEpochTime)
+    {
+        mDstOffsetObj.size = 0;
+        VerifyOrReturnValue(mTimeSyncDataProvider.ClearDSTOffset() == CHIP_NO_ERROR, TimeState::kInvalid);
+        // list is empty, generate DSTTableEmpty event
+        sendDSTTableEmptyEvent(GetDelegate()->GetEndpoint());
+        return TimeState::kInvalid;
+    }
+    if (activeDstIndex > 0)
+    {
+        mDstOffsetObj.size    = dstList.size() - activeDstIndex;
+        auto newDstOffsetList = dstList.SubSpan(activeDstIndex);
+        VerifyOrReturnValue(mTimeSyncDataProvider.StoreDSTOffset(newDstOffsetList) == CHIP_NO_ERROR, TimeState::kInvalid);
+        VerifyOrReturnValue(mTimeSyncDataProvider.LoadDSTOffset(mDstOffsetObj) == CHIP_NO_ERROR, TimeState::kInvalid);
+        return TimeState::kChanged;
+    }
+    return TimeState::kActive;
+}
+
+TimeSyncEventFlag TimeSynchronizationServer::GetEventFlag()
+{
+    return mEventFlag;
+}
+
+void TimeSynchronizationServer::ClearEventFlag(TimeSyncEventFlag flag)
+{
+    uint8_t eventFlag = to_underlying(mEventFlag) ^ to_underlying(flag);
```

Comment:
This doesn't make sense: if mEventFlag does not have the flag set, it will _set_ it.

Followup, but see comments above about removing this mEventFlag thing altogether.

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, resolving this comment for now, further followup on the issue.